File: DataRecovery/Ruby/data_recovery.rb
# data recovery solution in ruby for code eval by steven a dunn

input_file = ARGV[0]

File.open(input_file, 'r') do |file|
  while line = file.gets
    line = line.split(';')
    text = line[0].split
    indices = line[1].split.map!(&:to_i)

    missing_index = 0
    sorted_indices = indices.sort
    sorted_indices.each do |index|
      if index != missing_index + 1
        break
      end
      missing_index += 1
    end
    missing_index += 1

    mapping = text.zip(indices)
    mapping.each do |map| 
      if map[1].nil?
        map[1] = missing_index
      end
    end

    mapping.each do |map|
      map[0], map[1] = map[1], map[0]
    end
    mapping.sort!

    mapping.each do |map|
      print "#{map[1]} "
    end
    puts
  end
end

File: DataRecovery/c#/DataRecovery.cs
// data recovery soln in c# for code eval by steven a dunn

using System;
using System.IO;
using System.Linq;
using System.Collections.Generic;

class DataRecovery {
  public static int find_missing(List<int> indices, int length) {
    int sum = 0;
    foreach (int i in indices)
      sum += i;
    int expected_sum = length * (length + 1) / 2;
    return expected_sum - sum;
  }

  public static void Main(string[] args) {
    using (StreamReader reader = new StreamReader(args[0])) {
      while (!reader.EndOfStream) {
        string line = reader.ReadLine();
        if (null == line)
          continue;

        string[] inputs = line.Trim().Split(';');
        string[] words = inputs[0].Split();
        List<int> indices = inputs[1].Split().Select(x => Int32.Parse(x)).ToList();
        indices.Add(find_missing(indices, words.Length));
        for (int i = 1; i < indices.Count + 1; i++)
          Console.Write(words[indices.IndexOf(i)] + " ");
        Console.WriteLine();
      }
    }
  }
}